core/drivers/crypto/stm32/stm32_saes.c

```diff
@@ -1096,6 +1119,236 @@ TEE_Result stm32_saes_update(struct stm32_saes_context *ctx, bool last_block,
+		k2[sizeof(k2) - 1] = k2[sizeof(k2) - 1] ^ 0x87;
+
+	if (data_size > AES_BLOCK_SIZE) {
+		uint8_t *data_out = NULL;
+
+		/* All block but last in CBC mode */
+		res = stm32_saes_init(ctx, false, STM32_SAES_MODE_CBC,
+				      key_sel, key, key_size, block,
+				      sizeof(block));
+		if (res)
+			return res;
+
+		processed = ROUNDDOWN(data_size - 1, AES_BLOCK_SIZE);
+		data_out = malloc(processed);
+		if (!data_out)
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		res = stm32_saes_update(ctx, true, data, data_out, processed);
+		if (!res) {
+			/* Copy last out block or keep block as { 0 } */
+			memcpy(block, data_out + processed - AES_BLOCK_SIZE,
+			       AES_BLOCK_SIZE);
+		}
+
+		free(data_out);
+
+		if (res)
+			return res;
+	}
+
+	/* Manage last block */
+	xor_block(block, data + processed, data_size - processed);
+	if (data_size - processed == AES_BLOCK_SIZE) {
+		xor_block(block, k1, AES_BLOCK_SIZE);
+	} else {
+		/* xor with padding = 0b100... */
+		block[data_size - processed] ^= 0x80;
+		xor_block(block, k2, AES_BLOCK_SIZE);
+	}
+
+	/*
+	 * AES last block.
+	 * We need to use same chaining mode to keep same key if DHUK is
+	 * selected so we reuse l as a zero initialized IV.
+	 */
+	memset(l, 0, sizeof(l));
+	res = stm32_saes_init(ctx, false, STM32_SAES_MODE_CBC, key_sel, key,
+			      key_size, l, sizeof(l));
+	if (res)
+		return res;
+
+	return stm32_saes_update(ctx, true, block, out, AES_BLOCK_SIZE);
+}
+
+static TEE_Result stm32_saes_kdf(struct stm32_saes_context *ctx,
+				 enum stm32_saes_key_selection key_sel,
+				 const void *key, size_t key_size,
+				 const void *input, size_t input_size,
+				 uint8_t *subkey, size_t subkey_size)
+
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t index = 0;
+	uint32_t index_be = 0;
+	uint8_t *data = NULL;
+	size_t data_index = 0;
+	size_t subkey_index = 0;
+	size_t data_size = input_size + sizeof(index_be);
+	uint8_t cmac[AES_BLOCK_SIZE] = { };
+
+	if (!ctx || !input || !input_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* For each K(i) we will add an index */
+	data = malloc(data_size);
+	if (!data)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	data_index = 0;
+	index_be = TOBE32(index);
+	memcpy(data + data_index, &index_be, sizeof(index_be));
+	data_index += sizeof(index_be);
+	memcpy(data + data_index, input, input_size);
+	data_index += input_size;
+
+	/* K(i) computation. */
+	index = 0;
+	while (subkey_index < subkey_size) {
+		index++;
+		index_be = TOBE32(index);
+		memcpy(data, &index_be, sizeof(index_be));
+
+		res = stm32_saes_cmac_prf_128(ctx, key_sel, key, key_size,
+					      data, data_size, cmac);
+		if (res)
+			goto out;
+
+		memcpy(subkey + subkey_index, cmac,
+		       MIN(subkey_size - subkey_index, sizeof(cmac)));
+		subkey_index += sizeof(cmac);
+	}
+
+out:
+	free(data);
+	if (res)
+		memzero_explicit(subkey, subkey_size);
+
+	return res;
+}
+
+/* Implement hardware HUK derivation using SAES resources */
+TEE_Result huk_subkey_derive(enum huk_subkey_usage usage,
```

Comment:
I'd expect this to be compatible with the default implementation, but it looks like you get different results depending on implementation. These subkeys are in a way part of different ABIs provided by OP-TEE, if they change depending on the configuration you also break the ABIs.

Reply:
Hello,
I'm not sure to fully understand your comment, for you the HW HUK derivation should be the same algorithm and produce the same DHUK as the SW implementation in OPTEE ?

If yes, I don't think it's possible. The SAES does not support HMAC operation and the SW implementation uses HMAC.
Why is it so important that several configurations of OP-TEE derive the HUK the same way ? If a developer choose to change the derivation algorithm the secure storage and any other services must be reinitialised to use the new DHUKs produce with the new algorithm. Data encoded with the DHUK from the previous algorithm are not relevant anymore.

Reply:
If you're not going to give the same result as the SW version, then you shouldn't fall back to it either or it may lead to a very bad user experience.

Reply:
Ok I see your point.
On STM32MP13 the HUK is only accessible by SAES and cannot be read by the SW, therefore the key derivation has to be done with the HW.
By default, we compile all the drivers compatible for the platform and the configuration in the DT probes or not the drivers. 
For SAES, compiling the driver leads to the redefinition of the HUK derivation algorithm. If the peripheral is disabled in the DT then when we will derive a key we will be using an unprobbed peripheral and OP-TEE will panic. Hence the default implementation if SAES is disabled. (without the SW fall back, we would need to know at compile time if SAES will be enabled or not which is a really bad user experience)

Maybe, I try to fix the issue the wrong way and a weak function is not the right solution. 
We could add an ops that by default is pointing to the SW derivation provided by OP-TEE and when we probe SAES we change the ops to point to the SAES HW implementation.
What do you think ? Do you have a other propositions ?

Reply:
Where do you keep the DTB, compiled into the OP-TEE binary? If so, you know at compile-time whether SAES will be enabled. As long as there's no impact on generic code I'm not too concerned about what you're doing with plat-stm*. But when you need generic changes based on somewhat dubious reasons there's a problem. Having `huk_subkey_derive()` weak is OK, but adding a generic wrapper function too means forcing everyone else to use an extra indirection for no good reason.

Reply:
For now yes the DTB is in the OP-TEE binary but in the long term we will externalize the DTB to avoid recompiling the core when we want to change the configuration. So Parsing the DTB during the compilation to see if SAES is enabled or not is a no go for us.
And even now, we don't want to change the compilation of OP-TEE depending of the device tree. The code compiled should work independently of the status of SAES.

Yes there is an extra indirection but GCC will most probably optimize this, so I don't think it has an impact on the performance. And in term of support I suppose it will be similar to  `crypto_acipher_rsaes_encrypt()`.

So there are 2 good reasons to had the indirection : 
- Avoid a custom command line if we disable the HW that implements the HUK derivation.
- Avoid a panic if the HUK derivation is implemented with a disabled HW in the DT.